### ISSUE TITLE
Deferred.fromCoroutine has a redundant test against GeneratorType

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10329]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#10329|
|Type|enhancement|
|Created|2022-04-11 15:12:11Z|

`fromCoroutine` is implemented like:

```
# type note: Subclass of "Generator[Deferred[_T], object, _T]" and "GeneratorType" cannot exist                                                                           
if not iscoroutine(coro) and not isinstance(coro, GeneratorType):
    raise NotACoroutineError(f"{coro!r} is not a coroutine")

return _cancellableInlineCallbacks(coro)
```

`iscoroutine` is `asyncio.iscoroutine` and is implemented like:

```
# Prioritize native coroutine check to speed-up                                                                                                                                   
# asyncio.iscoroutine.                                                                                                                                                            
_COROUTINE_TYPES = (types.CoroutineType, types.GeneratorType,
                    collections.abc.Coroutine, CoroWrapper)
_iscoroutine_typecache = set()


def iscoroutine(obj):
    """Return True if obj is a coroutine object."""
    if type(obj) in _iscoroutine_typecache:
        return True

    if isinstance(obj, _COROUTINE_TYPES):
        # Just in case we don't want to cache more than 100                                                                                                                       
        # positive types.  That shouldn't ever happen, unless                                                                                                                     
        # someone stressing the system on purpose.                                                                                                                                
        if len(_iscoroutine_typecache) < 100:
            _iscoroutine_typecache.add(type(obj))
        return True
    else:
        return False
```

It seems like this makes the check in `fromCoroutine` equivalent to:

```
a = isinstance(coro, (types.CoroutineType, types.GeneratorType, collections.abc.Coroutine, CoroWrapper))
b = isinstance(coro, types.GeneratorType)
if not a and not b:
    raise NotACoroutineError(f"{coro!r} is not a coroutine")

return _cancellableInlineCallbacks(coro)
```

which is equivalent to:

```
a = isinstance(coro, (types.CoroutineType, types.GeneratorType, collections.abc.Coroutine, CoroWrapper))
if not a:
    raise NotACoroutineError(f"{coro!r} is not a coroutine")

return _cancellableInlineCallbacks(coro)
```

So, the `not isinstance(coro, GeneratorType)` check in `fromCoroutine` is redundant and can be removed.

<details><summary>Searchable metadata</summary>

```
trac-id__10329 10329
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1649689931557284 1649689931557284
changetime__1649724987954791 1649724987954791
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

https://github.com/twisted/twisted/pull/1717

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set status to `closed`|
|-|-|

Merged in c67cadf3c01e0a4c46862f9fbfaca42e96a46393
